### PR TITLE
Initial draft for data mover design

### LINKED ISSUE #4112: Design doc for data movement layer
Write a design doc for the data movement layer described in #3229

### PR DESCRIPTION
**Pull request by zubron**
# Please add a summary of your change
This contains the high-level design for the data mover.
The detailed design is still in progress and this PR will be updated further to include this.

Signed-off-by: Bridget McErlean <bmcerlean@vmware.com>

# Does your change fix a particular issue?

Fixes #4112 

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: design/data-mover.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+The Astrolabe API can be implemented for resources other than volumes, for example, databases.
+This enables each resource to define the best way to back up that entity and any subresources that it depends on.
+For example, when backing up a Postgres database, it may be preferable to use pg_dump rather than quiescing and snapshotting the volumes.
+This can be achieved using Astrolabe and as a result will provide data readers for the dump file.
+This data reader can be used to write the dump contents to an arbitrary backup location, rather than being confined to the underlying storage system.
+
+By allowing Velero to interact with Astrolabe Protected Entities, we can utilize the data readers for these entities to read the data and write to other locations.
+
+## Goals
+- Enable the data from snapshots of resources to be moved to multiple storage locations including object stores.
```
Comment (reasonerjt) [inline design/data-mover.md:32]: could you elaborate will it support moving data to storage locations other than object store?  Is it supported in velero now, and does it map to BSL?

File: design/data-mover.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+- Allow the location for the data store to be specified at backup time and allow selection of backup location at restore time.
+- Allow all or a subset of Protected Entities to be copied into the backup location.
```
Comment (reasonerjt) [inline design/data-mover.md:34]: Currently, there are no `Protected Entities` in velero.  Will implementing the data mover require implementing `Protected Entities`  in velero?

File: design/data-mover.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+This design does not consider removing existing VolumeSnapshot or restic capabilities. The VolumeSnapshot plugin is deprecated in favour of the ItemSnapshot plugin and restic will likely be replaced with Kopia, however support for these will not be removed as part of this work.
+
+## High-level Design
+Adding data movement to Velero is a significant feature.
+Due to this, the design is being broken into a number of phases where each phase provides benefit and functionality. 
+The data movement will be part of the Velero codebase, but later it may be moved to Astrolabe to allow Protected Entities to call it or use it as a destination.
+
+### Phase 1
+The movement of data will be triggered and controlled by Velero as part of the backup and restore process.
+Once the snapshotting phase of the backup is complete, Velero will iterate over ItemSnapshots (resource snapshotted using an ItemSnapshotter plugin), and retrieve the Astrolabe Protected Entities for these resources.
```
Comment (reasonerjt) [inline design/data-mover.md:51]: So does it implies that a prerequisite is to implement ItemSnapshotter that can take snapshot to user's workloads?

For example we migrate the volumesnap shotter to ItemSnapshotter first and then we may start to implement the data mover?

**Comment by reasonerjt** [comment]
I wish to clarify whether the implementation of the data mover depends on implementation of ItemSnapshotter plugin and `Protected Entity`

File: design/data-mover.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+Velero also has beta support for taking snapshots using CSI.
+Without data movement though, there is no guarantee that the snapshots are durable as it is dependent on the storage provider.
+
+Extraction of data from snapshots is something that is already available in the [Velero Plugin for vSphere](https://github.com/vmware-tanzu/velero-plugin-for-vsphere).
+This plugin makes use of the Astrolabe APIs to implement a vSphere FCD Protected Entity.
+This Protected Entity interface provides data readers and writers for the disks.
+Once a snapshot is created, the data reader is used to read the data from the disk and it is written to an S3 compatible object store.
+
+The Astrolabe API can be implemented for resources other than volumes, for example, databases.
```
Comment (reasonerjt) [inline design/data-mover.md:23]: Is this relevant to this design?  Will data mover only cover the data in `snapshot` or will we consider the dump file also a `snapshot`?

Currently, in velero the snapshot is the narrower concept.

File: design/data-mover.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+## Goals
+- Enable the data from snapshots of resources to be moved to multiple storage locations including object stores.
+- Allow the location for the data store to be specified at backup time and allow selection of backup location at restore time.
+- Allow all or a subset of Protected Entities to be copied into the backup location.
+- Allow the data mover functionality to be replaced by third parties.
+- Include performance considerations:
+  - Move data in a scale-out fashion
+  - Allow data movement to be throttled
+  - Support incremental backups
```
Comment (yanji09) [inline design/data-mover.md:39]: Does 'support incremental backup' means support to move incremental backup data? or mean data mover will provide similar function as CBT?

File: design/data-mover.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
+# Velero Data Mover
+
+## Abstract
+Velero currently does not handle data movement of snapshotted data.  We would like to have the ability to move data that has been snapshotted in a storage system or other [Astrolabe Protected Entity](https://github.com/vmware-tanzu/astrolabe) provider into the backup storage location specified.
+
+## Background
+Velero has two options for backing up persistent volumes:
+Using snapshotting APIs provided by the storage provider (for example, using the VolumeSnapshotter)
+Filesystem level backup using restic
```
Comment (jglick) [inline design/data-mover.md:9]:
```suggestion
* Using snapshotting APIs provided by the storage provider (for example, using the VolumeSnapshotter)
* Filesystem level backup using restic
```

File: design/data-mover.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+This enables each resource to define the best way to back up that entity and any subresources that it depends on.
+For example, when backing up a Postgres database, it may be preferable to use pg_dump rather than quiescing and snapshotting the volumes.
+This can be achieved using Astrolabe and as a result will provide data readers for the dump file.
+This data reader can be used to write the dump contents to an arbitrary backup location, rather than being confined to the underlying storage system.
+
+By allowing Velero to interact with Astrolabe Protected Entities, we can utilize the data readers for these entities to read the data and write to other locations.
+
+## Goals
+- Enable the data from snapshots of resources to be moved to multiple storage locations including object stores.
+- Allow the location for the data store to be specified at backup time and allow selection of backup location at restore time.
```
Comment (jglick) [inline design/data-mover.md:33]: To clarify: does this include replication at backup time? For example https://github.com/vmware-tanzu/velero-plugin-for-aws/pull/90 allows you to create an EBS snapshot in one region and then replicate it to another, which means two snapshot ids must be saved in backup metadata; the appropriate id is selected at restore time according to the region in which the restore is running. In that example the backup storage location for metadata just points to one region (since you can access the blob globally and performance is not an issue).

**Comment by reasonerjt** [comment]
Per discussion this design may need some rework and @zubron is leaving the team.
I'm closing this one and we will open a PR when a new design is ready after more discussions and thinking.
